File: labs/mobile/donor-archive/tunnel/MultiprotoEgressSelector.kt
package com.luminet.android.tunnel

data class AndroidEgressTarget(
    val targetId: String,
    val protocol: String,
    val weight: Int,
    var latencyMs: Int,
    var lossPercent: Double,
    var isActive: Boolean = true
)

class MultiprotoEgressSelector {
    private val targets = mutableMapOf<String, AndroidEgressTarget>()

    fun addTarget(target: AndroidEgressTarget) {
        targets[target.targetId] = target
    }

    fun updateMetrics(targetId: String, latencyMs: Int, lossPercent: Double, isActive: Boolean) {
        val t = targets[targetId] ?: return
        t.latencyMs = latencyMs
        t.lossPercent = lossPercent
        t.isActive = isActive
    }

    fun selectBest(preferredProto: String? = null): AndroidEgressTarget? {
        var best: AndroidEgressTarget? = null
        var bestScore = -1e9

        for (t in targets.values) {
            if (!t.isActive) continue
            var score = (t.weight * 10.0) - t.latencyMs - (t.lossPercent * 20.0)
            if (preferredProto != null && t.protocol == preferredProto) {
                score += 50.0
            }
            if (score > bestScore) {
                bestScore = score
                best = t
            }
        }
        return best
    }
}

File: labs/control-ui/api-shadow/multiproto_egress_selector.ts
export interface UiEgressTarget {
  targetId: string;
  protocol: string;
  weight: number;
  latencyMs: number;
  lossPercent: number;
  isActive?: boolean;
}

export class MultiprotoEgressSelector {
  private targets: Map<string, UiEgressTarget> = new Map();

  addTarget(target: UiEgressTarget): void {
    this.targets.set(target.targetId, { ...target, isActive: target.isActive ?? true });
  }

  updateMetrics(targetId: string, latencyMs: number, lossPercent: number, isActive: boolean): boolean {
    const t = this.targets.get(targetId);
    if (!t) return false;
    t.latencyMs = latencyMs;
    t.lossPercent = lossPercent;
    t.isActive = isActive;
    return true;
  }

  selectBest(preferredProto?: string): UiEgressTarget | null {
    let best: UiEgressTarget | null = null;
    let bestScore = -1e9;

    for (const t of this.targets.values()) {
      if (!t.isActive) continue;
      let score = t.weight * 10 - t.latencyMs - t.lossPercent * 20;
      if (preferredProto && t.protocol === preferredProto) {
        score += 50;
      }
      if (score > bestScore) {
        bestScore = score;
        best = t;
      }
    }

    return best;
  }
}
